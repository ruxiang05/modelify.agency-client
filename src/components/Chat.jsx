import React from 'react';
import Page from './Page';

const Chat = () => <Page>Chat</Page>;

export default Chat;
