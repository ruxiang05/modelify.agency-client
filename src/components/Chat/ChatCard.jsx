import React from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import '../../styles/chat-card.scss';
import { ReactComponent as UnknownUserIcon } from '../../assets/icons/user-circle.svg';
import { ReactComponent as GoTo } from '../../assets/icons/arrow-right.svg';

const ChatCard = ({ chat }) => (
  <Link
    to={{
      pathname: `/chats/${chat._id}`,
      state: {
        chat,
      },
    }}
  >
    <div className="chat-card">
      <div className="content">
        <UnknownUserIcon />
        <div>{chat.job.title}</div>

      </div>
      <GoTo />
    </div>
  </Link>
);

ChatCard.propTypes = {
  chat: PropTypes.shape({}).isRequired,
};
export default ChatCard;
