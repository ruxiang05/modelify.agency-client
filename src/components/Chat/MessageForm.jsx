/* Uses methods and/or components from react,prop-types and socket.io-client */
import React from 'react';
import PropTypes from 'prop-types';
import { ReactComponent as SendIcon } from '../../assets/icons/arrow-circle-right.svg';

class MessageForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      message: '',
    };

    this.sendMessage = this.sendMessage.bind(this);
    this.handleChange = this.handleChange.bind(this);
  }

  handleChange(event) {
    this.setState({ message: event.target.value });
  }

  sendMessage(event) {
    const { socket, user, chat } = this.props;
    const { message } = this.state;
    event.preventDefault();
    if (!message) return;
    socket.emit('send_message', {
      room: chat._id,
      user: user._id,
      message,
    });

    this.setState({ message: '' });
  }

  render() {
    const { message } = this.state;
    return (
      <div className="message-form">
        <input value={message} onChange={this.handleChange} />
        <button type="submit" onClick={this.sendMessage}><SendIcon /></button>
      </div>
    );
  }
}

MessageForm.propTypes = {
  user: PropTypes.shape({}).isRequired,
  socket: PropTypes.shape({}).isRequired,
  chat: PropTypes.shape({}).isRequired,
};

export default MessageForm;
