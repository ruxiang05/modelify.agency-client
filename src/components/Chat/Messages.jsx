/* Uses methods and/or components from react, prop-types, react-router-dom and socket.io-client */
import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

class Messages extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      messages: [],
    };

    this.messagesEnd = React.createRef();
  }

  componentDidMount() {
    const { socket, chat } = this.props;
    const { messages } = chat;
    if (messages) {
      this.setState({ messages });
    }

    this.scrollToEnd();

    socket.on('receive_message', (data) => {
      if (chat._id === data.room) {
        this.addMessage(data);
      }
    });
  }

  componentDidUpdate() {
    this.scrollToEnd();
  }

  addMessage(data) {
    this.setState(prevState => ({ messages: [...prevState.messages, data] }));
  }

  scrollToEnd() {
    this.messagesEnd.current.scrollIntoView({ behaviour: 'smooth' });
  }

  render() {
    const { messages } = this.state;
    const { chat, user } = this.props;
    const { job } = chat;
    return (
      <>
        <div className="job-chat-card">
          <h2>{job.title}</h2>
          <Link
            to={{
              pathname: `/jobs/${job._id}`,
              state: job,
            }}
          >
View

          </Link>
        </div>
        <div className="messages">
          {messages.map(message => (
            <div
              key={messages.indexOf(message)}
              className={message.user === user._id ? 'message sent' : 'message received'}
            >
              <p>
                {message.message}
              </p>
            </div>
          ))}
          <div ref={this.messagesEnd} />
        </div>
      </>
    );
  }
}

Messages.propTypes = {
  socket: PropTypes.shape({}).isRequired,
  user: PropTypes.shape({}).isRequired,
  chat: PropTypes.shape({}).isRequired,
};


export default Messages;
