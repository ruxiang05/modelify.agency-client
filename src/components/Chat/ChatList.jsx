import React from 'react';
import ChatCard from './ChatCard';
import PageHeader from '../PageHeader';
import api from '../../api';
import { getToken } from '../../auth';
import '../../styles/chat-list.scss';
import NoData from '../NoData';

class ChatList extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      chats: [],
    };
  }

  componentDidMount() {
    const token = getToken();
    api.chats.getChats(token).then((res) => {
      const { chats } = res;
      console.log(res);
      this.setState({ chats });
    });
  }

  render() {
    const { chats } = this.state;
    return (
      <React.Fragment>
        <PageHeader title="Chats" />
        {chats.length ? (
          <ul className="page chats">
            {chats.map(chat => (
              <li key={chats.indexOf(chat)}>
                <ChatCard chat={chat} />
              </li>
            ))}
          </ul>
        ) : (
          <NoData>
            <p>You do not have chats yet</p>
          </NoData>
        )}
      </React.Fragment>
    );
  }
}

export default ChatList;
