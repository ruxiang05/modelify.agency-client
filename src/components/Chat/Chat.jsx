/* Uses methods and/or components from react and prop-types */
import React from 'react';
import PropTypes from 'prop-types';
import { UserConsumer } from '../../contexts/userContext';
import { SocketConsumer } from '../../contexts/socketContext';
import MessageForm from './MessageForm';
import Messages from './Messages';
import PageHeader from '../PageHeader';
import '../../styles/chat.scss';
import api from '../../api';
import { getToken } from '../../auth';

class Chat extends React.Component {
  constructor(props) {
    super(props);

    const { location } = this.props;
    const { chat } = location.state;
    this.state = {
      chat,
    };
  }

  componentDidMount() {
    const { chat } = this.state;
    api.chats.getMessages(getToken(), chat._id).then((res) => {
      const { messages } = res;
      this.setState({ chat: { ...chat, messages } });
    });
  }

  render() {
    const { history } = this.props;
    const { chat } = this.state;
    const { users } = chat;
    const chatWith = (user => users.filter(u => u._id !== user._id)[0]);
    return (
      <UserConsumer>
        {({ user }) => (
          <SocketConsumer>
            {({ socket }) => (
              <>
                <PageHeader backButton history={history} title={chatWith(user).name} />
                <div className="page">
                  <Messages socket={socket} user={user} chat={chat} />
                  <MessageForm user={user} socket={socket} chat={chat} />
                </div>
              </>
            )}
          </SocketConsumer>
        )}
      </UserConsumer>
    );
  }
}

Chat.propTypes = {
  location: PropTypes.shape({}).isRequired,
  history: PropTypes.shape({}).isRequired,
};

export default Chat;
