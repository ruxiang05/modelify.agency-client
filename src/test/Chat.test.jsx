import React from 'react';
import { shallow } from 'enzyme';
import { createBrowserHistory } from 'history';
import Chat from '../components/Chat/Chat';

jest.mock('../api.js');

const agent = {
  id: '1',
  name: 'agent',
};

const model = {
  id: '2',
  name: 'model',
};

const chat = {
  id: '1',
  job: {
    title: 'title',
  },
  messages: [],
  users: [agent, model],
};

const location = {
  state: {
    chat,
  },
};

const history = createBrowserHistory({
  basename: '/',
  keyLength: 0,
});

const context = {
  user: agent,
  socket: {},
};

let wrapper;

beforeEach(() => {
  wrapper = shallow(<Chat location={location} history={history} />, {
    context,
  });
});

describe('Chat', () => {
  it('should render Chat', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
