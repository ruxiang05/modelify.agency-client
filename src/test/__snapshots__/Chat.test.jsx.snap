// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat should render Chat 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Chat
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
    location={
      Object {
        "state": Object {
          "chat": Object {
            "id": "1",
            "job": Object {
              "title": "title",
            },
            "messages": Array [],
            "users": Array [
              Object {
                "id": "1",
                "name": "agent",
              },
              Object {
                "id": "2",
                "name": "model",
              },
            ],
          },
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": [Function],
    },
    "ref": null,
    "rendered": [Function],
    "type": Object {
      "$$typeof": Symbol(react.context),
      "_calculateChangedBits": null,
      "_context": Object {
        "$$typeof": Symbol(react.context),
        "Consumer": [Circular],
        "Provider": Object {
          "$$typeof": Symbol(react.provider),
          "_context": [Circular],
        },
        "_calculateChangedBits": null,
        "_currentRenderer": null,
        "_currentRenderer2": null,
        "_currentValue": Object {
          "updateUser": [Function],
          "user": Object {},
        },
        "_currentValue2": Object {
          "updateUser": [Function],
          "user": Object {},
        },
        "_threadCount": 0,
      },
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": [Function],
      },
      "ref": null,
      "rendered": [Function],
      "type": Object {
        "$$typeof": Symbol(react.context),
        "_calculateChangedBits": null,
        "_context": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": [Circular],
          "Provider": Object {
            "$$typeof": Symbol(react.provider),
            "_context": [Circular],
          },
          "_calculateChangedBits": null,
          "_currentRenderer": null,
          "_currentRenderer2": null,
          "_currentValue": Object {
            "updateUser": [Function],
            "user": Object {},
          },
          "_currentValue2": Object {
            "updateUser": [Function],
            "user": Object {},
          },
          "_threadCount": 0,
        },
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "context": Object {
      "socket": Object {},
      "user": Object {
        "id": "1",
        "name": "agent",
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
